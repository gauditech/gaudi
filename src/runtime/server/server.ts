import express, { json } from "express";

import { setupEndpoints } from "@src/runtime/server/endpoints";
import { requestLogger, errorHandler } from "@src/runtime/server/middleware";
import { Definition } from "@src/types/definition";

export type CreateServerConfig = {
  port: number;
  host: string;
  definition: Definition;
};

export function createServer(config: CreateServerConfig) {
  const app = express();

  const port = config.port || 3000;
  const host = config.host || "0.0.0.0";

  app.use(json()); // middleware for parsing application/json body
  app.use(requestLogger);

  setupEndpoints(app, config.definition);

  app.use(errorHandler);

  app.listen(config.port, config.host, () => {
    console.log(`App is started on ${host}:${port}`);
  });

  return app;
}
