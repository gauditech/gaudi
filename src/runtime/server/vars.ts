import _ from "lodash";

export class Vars {
  _vars: Record<string, any>;
  constructor(params = {}) {
    this._vars = params;
  }
  set(name: string, value: any) {
    Object.assign(this._vars, { [name]: value });
  }
  get(name: string, path?: string[]): any {
    _.get(this._vars, [name, ...(path ?? [])]);
  }
  all(): any {
    return this._vars;
  }
  collect(path: string[]): any {
    collect(this._vars, path);
  }
  copy() {
    return new Vars(_.cloneDeep(this._vars));
  }
}

function collect(vars: Record<string, any>, path: string[]): any {
  if (_.isEmpty(path)) {
    return vars;
  }
  const [name, ...rest] = path;
  if (_.isArray(vars)) {
    return vars.flatMap((v) => collect(v[name], rest));
  } else {
    return collect(vars[name], rest);
  }
}
