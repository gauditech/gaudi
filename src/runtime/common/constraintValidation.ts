import _ from "lodash";

import { queryFromParts } from "../query/build.js";
import { executeQuery } from "../query/exec.js";
import { DbConn } from "../server/dbConn.js";
import { Vars } from "../server/vars.js";

import { getRef } from "@src/common/refs.js";
import {
  ActionDef,
  Definition,
  FieldsetDef,
  FieldsetFieldDef,
  TypedExprDef,
} from "@src/types/definition.js";

export type ReferenceIdResult = ValidReferenceIdResult | InvalidReferenceIdResult;
export type ValidReferenceIdResult = { fieldsetAccess: string[]; value: number };
export type InvalidReferenceIdResult = { fieldsetAccess: string[]; value: "no-reference" };

export async function fetchReferenceIds(
  def: Definition,
  dbConn: DbConn,
  actions: ActionDef[],
  input: Record<string, unknown>
): Promise<ReferenceIdResult[]> {
  const referenceInputs = actions.flatMap((action) => {
    if (action.kind !== "create-one" && action.kind !== "update-one") return [];
    return action.changeset.flatMap(({ name, setter }) => {
      if (setter.kind !== "fieldset-reference-input") return [];
      return [[name, setter] as const];
    });
  });

  const promiseEntries = referenceInputs.map(async ([_name, setter]) => {
    const field = getRef.field(def, setter.throughRefKey);

    const varName = field.name + "__input";
    const filter: TypedExprDef = {
      kind: "function",
      name: "is",
      args: [
        {
          kind: "alias",
          namePath: [field.modelRefKey, field.name],
        },
        {
          kind: "variable",
          name: varName,
          type: { kind: field.type, nullable: field.nullable },
        },
      ],
    };
    const queryName = field.modelRefKey + "." + field.name;
    const query = queryFromParts(def, queryName, [field.modelRefKey], filter, []);
    const inputValue = _.get(input, setter.fieldsetAccess);
    const result = await executeQuery(dbConn, def, query, new Vars({ [varName]: inputValue }), []);

    if (result.length === 0) {
      return { fieldsetAccess: setter.fieldsetAccess, value: "no-reference" as const };
    }
    if (result.length > 1) {
      throw Error(
        `Failed to find reference: There are multiple (${result.length}) '${field.modelRefKey}' where field '${field.name}' is '${inputValue}'`
      );
    }

    return { fieldsetAccess: setter.fieldsetAccess, value: result[0].id as number };
  });
  return Promise.all(promiseEntries);
}

/**
 * This mutates endpoint fieldset.
 * If reference has no id it will add "noReference" validator to fieldset field.
 */
export function assignNoReferenceValidators(
  fieldset: FieldsetDef,
  referenceIds: ReferenceIdResult[]
): asserts referenceIds is ValidReferenceIdResult[] {
  referenceIds.forEach((referenceIdValue) => {
    const { value, fieldsetAccess } = referenceIdValue;
    if (typeof value === "number") return;
    const currentFieldset = getNestedFieldset(fieldset, fieldsetAccess);
    currentFieldset.validators.push({ name: "noReference" });
  });
}

function getNestedFieldset(fieldset: FieldsetDef, fieldsetAccess: string[]): FieldsetFieldDef {
  if (fieldsetAccess.length === 0) {
    if (fieldset.kind === "record") {
      throw Error("Unexpected FieldsetRecordDef when searching for a nested fieldset");
    }
    return fieldset;
  }
  if (fieldset.kind === "field") {
    throw Error("Unexpected FieldsetFieldDef when searching for a nested fieldset");
  }
  const [head, ...tail] = fieldsetAccess;
  return getNestedFieldset(fieldset.record[head], tail);
}
