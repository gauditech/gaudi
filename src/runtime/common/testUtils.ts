import { compileToOldSpec } from "@src/compiler";
import { compose } from "@src/composer/composer";
import { QueryTree } from "@src/runtime/query/build";
import { NestedRow, QueryExecutor } from "@src/runtime/query/exec";
import { Vars } from "@src/runtime/server/vars";
import { CustomManyEndpointDef, Definition, FetchOneAction, QueryDef } from "@src/types/definition";

/**
 * Creates dummy query executor wich always return empty row.
 *
 * TODO: add abbility to define dummy return data
 */
export function mockQueryExecutor(): QueryExecutor {
  return {
    executeQuery(
      _def: Definition,
      _q: QueryDef,
      _params: Vars,
      _contextIds: number[]
    ): Promise<NestedRow[]> {
      return Promise.resolve([]);
    },

    executeQueryTree(
      _def: Definition,
      _qt: QueryTree,
      _params: Vars,
      _contextIds: number[]
    ): Promise<NestedRow[]> {
      return Promise.resolve([]);
    },
  };
}
/**
 * Provide models blueprint, and query blueprint separately.
 * This function will compile a blueprint and extract the query definition,
 * so you can use it in tests.
 */

export function makeTestQuery(models: string, query: string): { def: Definition; query: QueryDef } {
  const bp = `
  ${models}

  model TestHelperModel {}
  entrypoint TestHelperModel {
    custom endpoint {
      method GET
      cardinality many
      path "/test"
      action {
        fetch as q {
          ${query}
        }
      }
    }
  }
  `;
  const def = compose(compileToOldSpec(bp));
  const endpoint = def.entrypoints[0].endpoints[0] as CustomManyEndpointDef;
  const action = endpoint.actions[0] as FetchOneAction;
  return { def, query: action.query };
}
