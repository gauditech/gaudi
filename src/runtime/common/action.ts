import { applyFilterIdInContext, queryTreeFromParts, transformSelectPath } from "../query/build";
import { executeQueryTree } from "../query/exec";

import { dataToFieldDbnames, getRef } from "@src/common/refs";
import { assertUnreachable } from "@src/common/utils";
import { buildChangset as buildChangesetData } from "@src/runtime/common/changeset";
import { DbConn } from "@src/runtime/server/dbConn";
import { Vars } from "@src/runtime/server/vars";
import { ActionDef, CreateOneAction, Definition, UpdateOneAction } from "@src/types/definition";

export type ActionContext = {
  input: Record<string, unknown>;
  vars: Vars;
};

export async function executeActions(
  def: Definition,
  dbConn: DbConn,
  ctx: ActionContext,
  actions: ActionDef[]
) {
  for (const action of actions) {
    const model = getRef.model(def, action.model);
    const dbModel = model.dbname;

    const actionKind = action.kind;
    if (actionKind === "create-one") {
      const changesetData = buildChangesetData(action.changeset, ctx);
      const dbData = dataToFieldDbnames(model, changesetData);

      const id = await insertData(dbConn, dbModel, dbData);
      const deps = await fetchActionDeps(def, dbConn, action, id);
      deps && ctx.vars.set(action.alias, deps[0]);
    } else if (actionKind === "update-one") {
      const changesetData = buildChangesetData(action.changeset, ctx);
      const dbData = dataToFieldDbnames(model, changesetData);

      const targetId = resolveTargetId(ctx, action.targetPath);

      const id = await updateData(dbConn, dbModel, dbData, targetId);
      const deps = await fetchActionDeps(def, dbConn, action, id);
      deps && ctx.vars.set(action.alias, deps[0]);
    } else if (actionKind === "delete-one") {
      const targetId = resolveTargetId(ctx, action.targetPath);

      console.log("PRE DEL");

      await deleteData(dbConn, dbModel, targetId);
      console.log("POST DEL");
    } else {
      assertUnreachable(actionKind);
    }
  }
}

/**
 * Fetches the record targeted by action based on the actions `select` deps.
 */
async function fetchActionDeps(
  def: Definition,
  dbConn: DbConn,
  action: CreateOneAction | UpdateOneAction,
  id: number | null
): Promise<Record<string, unknown>[] | undefined> {
  if (!id) {
    throw new Error(`Failed to insert into ${action.model}`);
  }
  if (!action.alias) return;
  // no need to fetch if only ID is requested
  if (action.select.findIndex((item) => item.alias !== "id") < 0) {
    return [{ id }];
  }
  const qt = queryTreeFromParts(
    def,
    action.alias,
    [action.model],
    applyFilterIdInContext([action.model]),
    transformSelectPath(action.select, [action.alias], [action.model])
  );
  return executeQueryTree(dbConn, def, qt, new Vars(), [id]);
}

function resolveTargetId(ctx: ActionContext, targetPath: string[]): number {
  // append "id" to the end of the target path and separate it from it's root
  // eg.
  // [org, repo, issue] -> [org, repo, issue, id] -> [org, [repo, issue, id]]
  const [rootTargetName, ...path] = [...targetPath, "id"];

  return ctx.vars.get(rootTargetName, path);
}

// ---------- DB functions

async function updateData(
  dbConn: DbConn,
  model: string,
  data: Record<string, unknown>,
  targetId: number
): Promise<number | null> {
  // TODO: return action's `select` here
  const ret = await dbConn(model).update(data).where({ id: targetId }).returning("id");

  // FIXME findOne? handle unexpected result
  if (!ret.length) return null;
  return ret[0].id;
}

async function insertData(
  dbConn: DbConn,
  model: string,
  data: Record<string, unknown>
): Promise<number | null> {
  // TODO: return action's `select` here
  const ret = await dbConn.insert(data).into(model).returning("id");

  // FIXME findOne? handle unexpected result
  if (!ret.length) return null;
  return ret[0].id;
}

async function deleteData(dbConn: DbConn, model: string, targetId: number): Promise<number | null> {
  // TODO: return action's `select` here
  const ret = await dbConn(model).delete().where({ id: targetId }).returning("id");

  // FIXME findOne? handle unexpected result
  if (!ret.length) return null;
  return ret[0].id;
}
