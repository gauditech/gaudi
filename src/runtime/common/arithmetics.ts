import _ from "lodash";

import { assertUnreachable, ensureEqual } from "@src/common/utils";
import { FunctionName } from "@src/types/definition";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function fnNameToFunction(name: FunctionName): (...args: any[]) => unknown {
  switch (name) {
    case "+":
      return _.add;
    case "-":
      return _.subtract;
    case "*":
      return _.multiply;
    case "/":
      return _.divide;
    case "<":
      return _.lt;
    case ">":
      return _.gt;
    case "<=":
      return _.lte;
    case ">=":
      return _.gte;
    case "is":
      return _.isEqual;
    case "is not":
      return (a: unknown, b: unknown) => !_.isEqual(a, b);
    case "and":
      return (a: unknown, b: unknown) => a && b;
    case "or":
      return (a: unknown, b: unknown) => a || b;
    case "in":
      return _.includes;
    case "not in":
      return (a: unknown[], v: unknown) => !_.includes(a, v);
    case "concat":
      return (a: unknown[]) => a.join("");
    case "length":
      return (value: string) => value.length;
    case "lower":
      return (value: string) => value.toLowerCase();
    case "upper":
      return (value: string) => value.toUpperCase();
    default:
      return assertUnreachable(name);
  }
}

interface ITyped<T> {
  name: FunctionName;
  args: T[];
}

export async function executeArithmetics<T>(
  func: ITyped<T>,
  getValue: (name: T) => Promise<unknown>
): Promise<unknown> {
  switch (func.name) {
    case "+":
    case "-":
    case "*":
    case "/":
    case ">":
    case "<":
    case ">=":
    case "<=": {
      ensureEqual(func.args.length, 2);
      const val1 = await getValue(func.args[0]);
      const val2 = await getValue(func.args[1]);

      return fnNameToFunction(func.name)(val1, val2);
    }
    case "and":
    case "or": {
      ensureEqual(func.args.length, 2);
      const val1 = await getValue(func.args[0]);
      const val2 = await getValue(func.args[1]);

      return fnNameToFunction(func.name)(val1, val2);
    }
    case "is":
    case "is not":
    case "in":
    case "not in": {
      ensureEqual(func.args.length, 2);
      const val1 = await getValue(func.args[0]);
      const val2 = await getValue(func.args[1]);

      return fnNameToFunction(func.name)(val1, val2);
    }
    case "concat": {
      const vals = await Promise.all(
        func.args.map(async (arg) => {
          const value = await getValue(arg);

          return value;
        })
      );

      return fnNameToFunction(func.name)(vals);
    }
    case "length":
    case "lower":
    case "upper": {
      ensureEqual(func.args.length, 1);
      const val = await getValue(func.args[0]);

      return fnNameToFunction(func.name)(val);
    }
    default: {
      return assertUnreachable(func.name);
    }
  }
}
