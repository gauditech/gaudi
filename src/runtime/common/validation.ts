import {
  AnySchema,
  BaseSchema,
  BooleanSchema,
  NumberSchema,
  StringSchema,
  ValidationError,
  boolean,
  number,
  object,
  string,
} from "yup";

import { getHook } from "../hooks";

import { assertUnreachable } from "@src/common/utils";
import { BusinessError } from "@src/runtime/server/error";
import {
  FieldsetDef,
  FieldsetFieldDef,
  FieldsetRecordDef,
  HookValidator,
} from "@src/types/definition";

// ----- validation&transformation

export async function validateEndpointFieldset<R = Record<string, unknown>>(
  fieldset: FieldsetDef,
  data: Record<string, unknown>
): Promise<R> {
  try {
    const validationSchema = buildFieldsetValidationSchema(fieldset);
    return await validateRecord(data, validationSchema);
  } catch (err: unknown) {
    if (err instanceof ValidationError) {
      // error should be an instance of Yup's ValidationError
      // https://github.com/jquense/yup#validationerrorerrors-string--arraystring-value-any-path-string

      throw new BusinessError(
        "ERROR_CODE_VALIDATION",
        "Validation error",
        createRecordValidationError(err)
      );
    } else {
      throw err;
    }
  }
}

export type RecordValidationError = Pick<
  ValidationError,
  "value" | "path" | "type" | "errors" | "params" | "inner"
>;

export function createRecordValidationError(error: ValidationError): RecordValidationError {
  return {
    value: error.value,
    path: error.path,
    type: error.type,
    errors: error.errors,
    params: error.params,
    inner: error.inner,
  };
}

export async function validateRecord(record: unknown, schema: AnySchema) {
  return schema.validate(record, {
    abortEarly: false, // report ALL errors, not just the first one
  });
}

// ---------- fieldset validation builder

export function buildFieldsetValidationSchema(fieldset: FieldsetDef): AnySchema {
  if (fieldset.kind !== "record") throw new Error('Root fieldset must be of kind "record".');

  return buildObjectValidationSchema(fieldset);
}

function processFields(name: string, field: FieldsetDef): AnySchema {
  if (field.kind === "field") {
    return buildFieldValidationSchema(name, field);
  } else {
    return buildObjectValidationSchema(field);
  }
}

function buildObjectValidationSchema(field: FieldsetRecordDef): AnySchema {
  const fieldSchemas = Object.fromEntries(
    Object.entries(field.record).map(([name, value]) => [name, processFields(name, value)])
  );

  if (!field.nullable) object(fieldSchemas).required();

  return object(fieldSchemas).optional();
}

function buildFieldValidationSchema(name: string, field: FieldsetFieldDef): AnySchema {
  if (field.type === "text") {
    // start with nullable because it's the only way to
    let s = string();

    if (field.nullable) {
      // TODO: yup's types don't allow expanding return type to `string | undefined | null`
      s = s.nullable() as StringSchema;
    }
    if (field.required) {
      // everything except `undefined`
      s = s.defined();
    }

    field.validators.forEach((v) => {
      if (v.name === "minLength") {
        s = s.min(v.args[0].value);
      } else if (v.name === "maxLength") {
        s = s.max(v.args[0].value);
      } else if (v.name === "isEmail") {
        s = s.email();
      } else if (v.name === "isTextEqual") {
        // TODO: s.equals returns BaseSchema which doesn't fit StringSchema
        s = s.equals<string>([v.args[0].value]) as StringSchema;
      } else if (v.name === "hook") {
        s = buildHookSchema(name, v, s);
      }
    });

    return s;
  } else if (field.type === "integer") {
    let s = number();

    if (field.nullable) {
      // TODO: yup's types don't allow expanding return type to `number | undefined | null`
      s = s.nullable() as NumberSchema;
    }
    if (field.required) {
      // everything except `undefined`
      s = s.defined();
    }

    field.validators.forEach((v) => {
      if (v.name === "min") {
        s = s.min(v.args[0].value);
      } else if (v.name === "max") {
        s = s.max(v.args[0].value);
      } else if (v.name === "isIntEqual") {
        // TODO: s.equals returns BaseSchema which doesn't fit NumberSchema
        s = s.equals([v.args[0].value]) as NumberSchema;
      } else if (v.name === "hook") {
        s = buildHookSchema(name, v, s);
      }
    });

    return s;
  } else if (field.type === "boolean") {
    let s = boolean();

    // NOTE: boolean schema cannot be "nullable"

    if (field.required) {
      // everything except `undefined`
      s = s.defined();
    }

    field.validators.forEach((v) => {
      if (v.name === "isBoolEqual") {
        // TODO: s.equals returns BaseSchema which doesn't fit BooleanSchema
        s = s.equals([v.args[0].value]) as BooleanSchema;
      } else if (v.name === "hook") {
        s = buildHookSchema(name, v, s);
      }
    });

    return s;
  } else {
    assertUnreachable(field.type);
  }
}

function buildHookSchema<S extends BaseSchema>(
  name: string,
  validator: HookValidator,
  schema: S
): S {
  let testFn: (value: unknown) => boolean;
  const code = validator.code;
  const args = validator.args;
  const hasArg = args.find((arg) => arg.name === name) != null;

  if (code.kind === "inline") {
    testFn = (value) => {
      return eval(hasArg ? `const ${name} = ${JSON.stringify(value)};${code.inline}` : code.inline);
    };
  } else {
    testFn = (value) => {
      const hook = getHook(code.file, code.target);
      if (hasArg) {
        return hook(value);
      } else {
        return hook();
      }
    };
  }

  return schema.test(testFn);
}
