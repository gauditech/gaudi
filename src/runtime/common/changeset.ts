import _, { get, indexOf, isString, set, toInteger, toString } from "lodash";

import { executeArithmetics } from "./arithmetics";

import { assertUnreachable, ensureEqual, ensureNot } from "@src/common/utils";
import { ActionContext } from "@src/runtime/common/action";
import { executeHook } from "@src/runtime/hooks";
import { ChangesetDef, FieldDef, FieldSetter } from "@src/types/definition";

type Changeset = Record<string, unknown>;

/**
 * Build result record from given action changeset rules and give context (source) inputs.
 */
export async function buildChangeset(
  actionChangsetDefinition: ChangesetDef,
  actionContext: ActionContext,
  // `changesetContext` is used for hooks, to be able to pass the "parent context" changeset
  changesetContext: Changeset = {}
): Promise<Changeset> {
  const changeset: Changeset = {};

  async function getValue(setter: FieldSetter): Promise<unknown> {
    switch (setter.kind) {
      case "literal": {
        return formatFieldValue(setter.value, setter.type);
      }
      case "fieldset-virtual-input":
      case "fieldset-input": {
        return formatFieldValue(
          getFieldsetProperty(actionContext.input, setter.fieldsetAccess),
          setter.type
        );
      }
      case "reference-value": {
        return actionContext.vars.get(setter.target.alias, setter.target.access);
      }
      case "fieldset-hook": {
        const args = await buildChangeset(setter.args, actionContext, changeset);
        return await executeHook(setter.code, args);
      }
      case "changeset-reference": {
        /**
         * Composer guarantees the correct order of array elements, so `setter.referenceName` should
         * be in the context and we should be able to return `changeset[setter.referenceName]`.
         *
         * However, hooks inherit the action changeset context, but also build their own changeset,
         * so we need to check which changeset `setter.referenceName` belongs to. In other words,
         * it's possible that `setter.referenceName in changeset` is `false` when building a hooks changeset,
         * but in that case we can assume it's in the `changesetContext`.
         *
         * NOTE the following code wouldn't work because `undefined` is a valid changeset value:
         * `return changeset[setter.referenceName] || changesetContext[setter.referenceName];`
         */
        if (setter.referenceName in changeset) {
          return changeset[setter.referenceName];
        } else {
          ensureEqual(setter.referenceName in changesetContext, true);
          return changesetContext[setter.referenceName];
        }
      }
      case "fieldset-reference-input": {
        const referenceIdResult = actionContext.referenceIds.find((result) =>
          _.isEqual(result.fieldsetAccess, setter.fieldsetAccess)
        );
        ensureNot(referenceIdResult, undefined);
        return referenceIdResult.value;
      }
      case "function": {
        return executeArithmetics(setter, (s) => getValue(s));
      }
      case "context-reference":
        return actionContext.vars.get(setter.referenceName);

      default: {
        return assertUnreachable(setter);
      }
    }
  }

  for (const { name, setter } of actionChangsetDefinition) {
    switch (setter.kind) {
      case "fieldset-reference-input": {
        changeset[name + "_id"] = await getValue(setter);
        break;
      }
      default: {
        changeset[name] = await getValue(setter);
      }
    }
  }

  /**
   * Remove all the virtual fields' values from the changeset.
   */
  for (const { name, setter } of actionChangsetDefinition) {
    switch (setter.kind) {
      case "fieldset-virtual-input": {
        delete changeset[name];
      }
    }
  }

  return changeset;
}

/**
 * Format unknown field value to one of mapping types.
 *
 * If `value` is `undefined`/`null`, an `undefined` is returned.
 *
 * This function uses `lodash` to convert "text" and "integer".
 * For details see: https://lodash.com/docs
 *
 * Mappings:
 * - text - `string`, `_.toString()`
 * - integer - `number`, `_.toInteger()`
 * - boolean - `boolean`, see below
 *
 * Due to `lodash` lacking boolean converter, this fn does it manually following these rules:
 * - "true" - string "true", case insensitive ("true", "TRUE", "TruE", ...)
 * - true - real boolean true
 * - everything else is converted to `false`
 *
 * TODO: move to some utils folder if it's ok
 */
export function formatFieldValue(
  value: unknown,
  type: FieldDef["type"] | "null"
): string | number | boolean | undefined | null {
  if (_.isNil(value)) return value;

  if (type === "boolean") {
    if (isString(value)) {
      value = value.toLowerCase();
    }

    return indexOf(["true", true], value) != -1;
  } else if (type === "integer") {
    return toInteger(value);
  } else {
    return toString(value);
  }
}

// ----- transformations

export function getFieldsetProperty<T = unknown>(target: unknown, fieldsetAccess: string[]): T {
  return get(target, fieldsetAccessToPath(fieldsetAccess));
}
export function setFieldsetProperty(
  target: object,
  fieldsetAccess: string[],
  value: unknown
): unknown {
  set(target, fieldsetAccessToPath(fieldsetAccess), value);

  return target;
}

export function fieldsetAccessToPath(access: string[]): string {
  return access.join(".");
}
