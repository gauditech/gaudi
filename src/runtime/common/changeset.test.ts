import { compose } from "@src/composer/composer";
import { getTypedLiteralValue } from "@src/composer/utils";
import { ActionContext } from "@src/runtime/common/action";
import {
  buildChangeset,
  buildStrictChangeset,
  fieldsetAccessToPath,
  formatFieldValue,
  getFieldsetProperty,
  setFieldsetProperty,
} from "@src/runtime/common/changeset";
import { mockQueryExecutor } from "@src/runtime/common/testUtils";
import { Vars } from "@src/runtime/server/vars";
import {
  ChangesetDef,
  Definition,
  FieldSetter,
  FieldSetterChangesetReference,
  FieldSetterFunction,
  FunctionName,
} from "@src/types/definition";

describe("runtime", () => {
  describe("changeset", () => {
    it("build action changeset object", async () => {
      const data: ChangesetDef = [
        {
          name: "value_prop",
          setter: { kind: "literal", value: "just value", type: "text" },
        },
        {
          name: "input_prop",
          setter: {
            fieldsetAccess: ["input_prop"],
            kind: "fieldset-input",
            type: "text",
            required: true,
          },
        },
        {
          name: "input_prop_missing",
          setter: {
            fieldsetAccess: ["__missing__"],
            kind: "fieldset-input",
            type: "text",
            required: false,
          },
        },
        {
          name: "virtual_input_prop",
          setter: {
            fieldsetAccess: ["virtual_input_prop"],
            kind: "fieldset-virtual-input",
            type: "text",
            required: false,
            nullable: false,
            validators: [],
          },
        },
        {
          name: "input_value_copy",
          setter: {
            kind: "changeset-reference",
            referenceName: "input_prop",
          },
        },
        {
          name: "other_model",
          setter: {
            kind: "fieldset-reference-input",
            throughRefKey: "OtherModel.slug",
            fieldsetAccess: ["slug"],
          },
        },
        {
          name: "hook_field",
          setter: {
            kind: "fieldset-hook",
            args: [
              { name: "x", setter: { kind: "literal", type: "integer", value: 6 } },
              { name: "y", setter: { kind: "literal", type: "integer", value: 2 } },
            ],
            hook: {
              runtimeName: "TestRuntime",
              code: {
                kind: "inline",
                inline: "x / y",
              },
            },
          },
        },
      ];

      const context: ActionContext = {
        input: {
          input_prop: "input value",
          virtual_input_prop: "virtual input value",
        },
        vars: new Vars(),
        referenceIds: [{ fieldsetAccess: ["slug"], value: 1 }],
      };

      expect(
        await buildChangeset(createTestDefinition(), mockQueryExecutor(), data, context)
      ).toMatchSnapshot();
    });

    it("build strict action changeset object", async () => {
      const data: ChangesetDef = [
        // lept field
        {
          name: "input_prop",
          setter: { kind: "literal", value: "just value", type: "text" },
        },
        // removed virtual/transient fileds
        {
          name: "virtual_input_prop",
          setter: {
            fieldsetAccess: ["virtual_input_prop"],
            kind: "fieldset-virtual-input",
            type: "text",
            required: false,
            nullable: false,
            validators: [],
          },
        },
      ];

      const context: ActionContext = {
        input: {
          input_prop: "input value",
          virtual_input_prop: "virtual input value",
        },
        vars: new Vars(),
        referenceIds: [{ fieldsetAccess: ["slug"], value: 1 }],
      };

      expect(
        await buildStrictChangeset(createTestDefinition(), mockQueryExecutor(), data, context)
      ).toMatchSnapshot();
    });

    it("calculate changeset arithmetic operations", async () => {
      const mkRef = (referenceName: string): FieldSetterChangesetReference => ({
        kind: "changeset-reference",
        referenceName,
      });
      const mkFn = (name: FunctionName, args: FieldSetter[]): FieldSetterFunction => ({
        kind: "function",
        name,
        args,
      });

      const changeset: ChangesetDef = [
        { name: "a", setter: getTypedLiteralValue(2) },

        { name: "foo", setter: getTypedLiteralValue("foo1") },
        { name: "bar", setter: getTypedLiteralValue("bar2") },
        { name: "is_a", setter: getTypedLiteralValue(true) },

        { name: "plus", setter: mkFn("+", [mkRef("a"), getTypedLiteralValue(4)]) },
        { name: "minus", setter: mkFn("-", [mkRef("a"), getTypedLiteralValue(-3)]) },
        { name: "multiply", setter: mkFn("*", [getTypedLiteralValue(6), mkRef("a")]) },
        { name: "divide", setter: mkFn("/", [getTypedLiteralValue(6), mkRef("a")]) },

        { name: "gt", setter: mkFn(">", [mkRef("a"), getTypedLiteralValue(2)]) },
        { name: "gte", setter: mkFn(">=", [mkRef("a"), getTypedLiteralValue(2)]) },
        { name: "lt", setter: mkFn("<", [mkRef("a"), getTypedLiteralValue(2)]) },
        { name: "lte", setter: mkFn("<=", [mkRef("a"), getTypedLiteralValue(2)]) },

        { name: "and", setter: mkFn("and", [mkRef("is_a"), getTypedLiteralValue(false)]) },
        { name: "or", setter: mkFn("or", [mkRef("is_a"), getTypedLiteralValue(false)]) },

        { name: "is", setter: mkFn("is", [mkRef("a"), getTypedLiteralValue(4)]) },
        { name: "is not", setter: mkFn("is not", [mkRef("a"), getTypedLiteralValue(4)]) },

        { name: "in", setter: getTypedLiteralValue("TODO") },
        { name: "not in", setter: getTypedLiteralValue("TODO") },

        {
          name: "concat",
          setter: mkFn("concat", [mkRef("foo"), getTypedLiteralValue(" "), mkRef("bar")]),
        },
        { name: "length", setter: mkFn("length", [mkRef("foo")]) },
        { name: "lower", setter: mkFn("lower", [mkRef("foo")]) },
        { name: "upper", setter: mkFn("upper", [mkRef("foo")]) },
      ];
      const context: ActionContext = {
        input: {},
        vars: new Vars(),
        referenceIds: [],
      };
      expect(
        await buildChangeset(createTestDefinition(), mockQueryExecutor(), changeset, context)
      ).toMatchSnapshot();
    });
  });

  describe("transformations", () => {
    it("create property path from fieldset access", () => {
      expect(fieldsetAccessToPath([])).toEqual("");
      expect(fieldsetAccessToPath(["a", "b", "c"])).toEqual("a.b.c");
      expect(fieldsetAccessToPath(["a", "b", "0", "c"])).toEqual("a.b.0.c");
    });

    it("get object property from path", () => {
      const target = { a: { b: [{ c: 3 }], b1: 2 } };

      expect(getFieldsetProperty(target, ["a", "b", "0", "c"])).toEqual(3);
    });

    it("set object property to path", () => {
      const target = { a: { b: [{ c: 3 }] } };

      expect(setFieldsetProperty(target, ["a", "b", "0", "c"], 3)).toEqual({
        a: { b: [{ c: 3 }] },
      });
    });
  });

  describe("formatting", () => {
    it("ignores undefined/null vlaue", () => {
      expect(formatFieldValue(undefined, "text")).toStrictEqual(undefined);
      expect(formatFieldValue(null, "text")).toStrictEqual(null);
      expect(formatFieldValue(undefined, "null")).toStrictEqual(undefined);
      expect(formatFieldValue(null, "null")).toStrictEqual(null);
    });

    it("formats text field values", () => {
      expect(formatFieldValue("asdf", "text")).toStrictEqual("asdf");
      expect(formatFieldValue(123, "text")).toStrictEqual("123");
      expect(formatFieldValue([1, 2, 3], "text")).toStrictEqual("1,2,3");
      expect(formatFieldValue({ a: 1 }, "text")).toStrictEqual("[object Object]");
    });

    it("formats integer field values", () => {
      expect(formatFieldValue(123, "integer")).toStrictEqual(123);
      expect(formatFieldValue("asdf", "integer")).toStrictEqual(0);
      expect(formatFieldValue([1, 2, 3], "integer")).toStrictEqual(0);
      expect(formatFieldValue({ a: 1 }, "integer")).toStrictEqual(0);
    });

    it("formats boolean field values", () => {
      expect(formatFieldValue(true, "boolean")).toStrictEqual(true);
      expect(formatFieldValue("true", "boolean")).toStrictEqual(true);
      expect(formatFieldValue("TruE", "boolean")).toStrictEqual(true);
      expect(formatFieldValue("asdf", "boolean")).toStrictEqual(false);
      expect(formatFieldValue(false, "boolean")).toStrictEqual(false);
      expect(formatFieldValue("false", "boolean")).toStrictEqual(false);
      expect(formatFieldValue([1, 2, 3], "boolean")).toStrictEqual(false);
      expect(formatFieldValue({ a: 1 }, "boolean")).toStrictEqual(false);
    });
  });
});

/**
 * Creates dummy definition struct
 */
function createTestDefinition(): Definition {
  const def = compose({
    entrypoints: [],
    models: [],
    populators: [],
    runtimes: [
      {
        name: "TestRuntime",
        sourcePath: "./src/runtime/test/hooks",
      },
    ],
  });

  return def;
}
