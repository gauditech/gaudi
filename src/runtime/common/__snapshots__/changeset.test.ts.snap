// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runtime changeset build action changeset object 1`] = `
{
  "hook_field": 3,
  "input_prop": "input value",
  "input_prop_missing": undefined,
  "input_value_copy": "input value",
  "other_model_id": 1,
  "value_prop": "just value",
  "virtual_input_prop": "virtual input value",
}
`;

exports[`runtime changeset build strict action changeset object 1`] = `
{
  "input_prop": "just value",
}
`;

exports[`runtime changeset calculate changeset arithmetic operations 1`] = `
{
  "a": 2,
  "and": false,
  "bar": "bar2",
  "concat": "foo1 bar2",
  "divide": 3,
  "foo": "foo1",
  "gt": false,
  "gte": true,
  "in": "TODO",
  "is": false,
  "is not": true,
  "is_a": true,
  "length": 4,
  "lower": "foo1",
  "lt": false,
  "lte": true,
  "minus": 5,
  "multiply": 12,
  "not in": "TODO",
  "or": true,
  "plus": 6,
  "upper": "FOO1",
}
`;
